src/types/index.ts

```diff
@@ -7,10 +7,13 @@ export * from './BulkUpload';
 export * from './Bundle';
 export * from './Changemaker';
 export * from './ChangemakerProposal';
+export * from './ChangemakerRole';
 export * from './CheckResult';
 export * from './DataProvider';
+export * from './DataProviderRole';
 export * from './express/AuthenticatedRequest';
 export * from './Funder';
+export * from './FunderRole';
```

Comment:
I agree and for what it's worth I did explore this and implemented both ways locally to try it out -- the alternative unfortunately felt similarly awkward but also in a way that was limiting for future functionality.

**First the awkward;** A single table would mean either:

(1) Using polymorphic association: No foreign key relationships, but we can have a natural primary key
(2) Using null-able columns (what we do for `sources`): Foreign key relationships, but no natural primary keys

If we want both foreign keys and natural primary keys, having the three tables to represent the three types of role the only solution I came up with :-/

**Second, the the functionality / data consistency;** By having three tables for the three types of role relationship we can eventually change the `access_type` for a given org entity to be more granular and specific to the activities related to that organization.

For instance, we expect to have `edit_opportunities` and `edit_proposals` as separate access types in the context of a funder.  I was planning on building that in in a future iteration rather than here, but it doesn't feel like a stretch to consider now.

---

Not sure if that's at all compelling, but I think I at least reflected the exploration and the tradeoffs I found.

Reply:
I think I'm OK with the three tables as long as they are access/permissions tables (associated with objects) rather than roles tables (implying subjects).